### ISSUE TITLE
Calico does not work properly on systems with kernel version 4.x+ unless ipv6 network is disabled

### ISSUE THREAD

**Issue by ansiz** [projectcalico/calico#2191]
When I run:
```
docker run --privileged -tid --rm --network net2 --name k530-net2 harbor.hpc.com/images/busybox
```
docker reported a problem:
```
15ba23b49172c9dc4f0643f3f11984ce02c878a60bafccb268becec600330a8f
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: 
starting container process caused "process_linux.go:402: container init caused 
\"process_linux.go:385: running prestart hook 0 caused \\\"error running hook: exit status 1,
 stdout: , stderr: time=\\\\\\\"2018-09-16T22:25:13-04:00\\\\\\\" level=fatal msg=
\\\\\\\"failed to add interface temp31556e7d316 to sandbox: error setting interface 
\\\\\\\\\\\\\\\"temp31556e7d316\\\\\\\\\\\\\\\" routes to [\\\\\\\\\\\\\\\"169.254.1.1/32\\\\\\\\\\\\\\\" 
\\\\\\\\\\\\\\\"fe80::b448:31ff:fee4:de7d/128\\\\\\\\\\\\\\\"]: permission denied\\\\\\\"\\\\n\\\"\"": unknown.
```

I can run this command on standard CentOS 7.x with kernel 3.x and it also not work on ubuntu 18.04 which has kernel 4.x, I found some log in `dmesg`:

```
[ 2111.674564] IPv6: ADDRCONF(NETDEV_UP): temp66aa9bddf71: link is not ready
[ 2111.674700] IPv6: ADDRCONF(NETDEV_UP): cali66aa9bddf71: link is not ready
[ 2111.674710] IPv6: ADDRCONF(NETDEV_CHANGE): cali66aa9bddf71: link becomes ready
[ 2111.674760] IPv6: ADDRCONF(NETDEV_CHANGE): temp66aa9bddf71: link becomes ready
[ 2111.926941] cali0: renamed from temp66aa9bddf71
[ 2113.110629] IPv6: ADDRCONF(NETDEV_UP): tempf1169b462ad: link is not ready
[ 2113.111066] IPv6: ADDRCONF(NETDEV_CHANGE): tempf1169b462ad: link becomes ready
[ 2113.325654] cali0: renamed from tempf1169b462ad
[ 2114.395699] IPv6: ADDRCONF(NETDEV_UP): tempc99fe2a39dc: link is not ready
[ 2114.400374] IPv6: ADDRCONF(NETDEV_CHANGE): tempc99fe2a39dc: link becomes ready
[ 2114.571455] cali0: renamed from tempc99fe2a39dc
[ 2115.557923] IPv6: ADDRCONF(NETDEV_UP): tempa2528b66f07: link is not ready
[ 2115.563399] IPv6: ADDRCONF(NETDEV_CHANGE): tempa2528b66f07: link becomes ready
[ 2115.744184] cali0: renamed from tempa2528b66f07
```

So I try to disable ipv6 with command:
```
echo 1 > /proc/sys/net/ipv6/conf/default/disable_ipv6
echo 1 > /proc/sys/net/ipv6/conf/all/disable_ipv6
```

Then it works fine

## Expected Behavior

I hope Calico 2.6 can work properly on systems with kernel version 4.x without ipv6 disabled.

## Possible Solution

Disable ipv6
```
echo 1 > /proc/sys/net/ipv6/conf/default/disable_ipv6
echo 1 > /proc/sys/net/ipv6/conf/all/disable_ipv6
```

## Steps to Reproduce (for bugs)

1. Install Calico 2.6 on the systems with kernel 4.x+
2. Try to create a container with calico network

## Your Environment

* Calicoctl version v1.6.4, build ae98f46f
* Docker without orchestration
* Operating System and version: CentOS Linux release 7.5.1804 (Core) Kernel: Linux 4.18.7

**Comment by ti-mo** [comment]
Hi, thanks for the report! We're experiencing the exact same issue, but the behaviour seems very flakey. This started appearing for us when we went from 4.15.15 to 4.16.x (and now 4.18.10).

We are using libnetwork, which is likely the case for OP as well. Calico tries to set an IPv6 address on a container interface that should not be v6-enabled. Logging into the container does not show a (stateless) link-local fe80 or anything auto-assigned by the kernel. The Docker network's `EnableIPv6` is set to `False`, and none of the containers we run have anything set in their IPv6Address fields.

```
# docker network inspect <net>
...
                "IPv4Address": "10.123.121.83/32",
                "IPv6Address": ""
```

Could this be a fallback mechanism in case IPv6Address is empty? Newer kernel versions likely reject unwanted addresses instead of silently dropping the Netlink messages, or are rejected using a different errno.

@caseydavenport @fasaxc Any ideas?

**Comment by ansiz** [comment]
@ti-mo 
> We're experiencing the exact same issue, but the behaviour seems very flakey. Given many retries/re-schedules, chances are most containers will be successfully started eventually.

Yes, container will be successfully started after many retries, but the network cannot communicate even if the container is already started.

The same behavior with the command: `docker network connect`, the network cannot communicate even if the IP has allocated to container

**Comment by caseydavenport** [comment]
This sounds to me like the libnetwork-plugin is trying to assign an IPv6 address when it shouldn't.

It seems to decide how to do that here: https://github.com/projectcalico/libnetwork-plugin/blob/e9d4f6cb286beee23503a0aae8963bef5c0a84ea/driver/network_driver.go#L498-L509

Based off of whether or not an IPv6 LL address is available on the host. Maybe we want to make that configurable, or smarter in some way?

**Comment by ti-mo** [comment]
@caseydavenport That's indeed what I initially thought. This can only really work properly when libnetwork-plugin can query whether or not IPv6 is enabled on the target network. The Docker network in question has `"EnableIPv6": false,`, set when running `inspect` on it, because we don't explicitly enable this when creating our networks (as intended).

There's also the case of IPv6 being enabled on the Docker network, but sysctl disabled on the system, though this shouldn't cause problems because it will still cause `linkLocalAddr` to be `nil`.

Any ideas how we can query `EnableIPv6` in the target network?

**Comment by ti-mo** [comment]
@caseydavenport That's indeed what I initially thought. This can only really work properly when libnetwork-plugin can query whether or not IPv6 is enabled on the target network. The Docker network in question has `"EnableIPv6": false,`, set when running `inspect` on it, because we don't explicitly enable this when creating our networks (as intended).

There's also the case of IPv6 being enabled on the Docker network, but sysctl disabled on the system, though this shouldn't cause problems because it will still cause `linkLocalAddr` to be `nil`.

Any ideas how we can query `EnableIPv6` in the target network?

**Comment by caseydavenport** [comment]
> Any ideas how we can query EnableIPv6 in the target network?

Looks like we have some logic already to inspect the network, might be as simple as using something like this? https://github.com/projectcalico/libnetwork-plugin/blob/e9d4f6cb286beee23503a0aae8963bef5c0a84ea/driver/network_driver.go#L583-L590

**Comment by xichengqian** [comment]
OMG!
I just remeber it.
I justsolved it through your method.
Thank u very much!
